src/arithmetic/algebraic_expression.c

```diff
@@ -674,6 +676,53 @@ void AlgebraicExpression_Free(AlgebraicExpression *ae) {
 	free(ae);
 }
 
+/* Refetch all operands from graph.
+ * This is to guarantee operands destination comply with
+ * multiplication requirements.
+ * As algebraic expressions are composed prior to the actual query evaluation
+ * operands dimensions might be outdated as the query evaluates
+ * consider WRITE READ WRITE query:
+ * MERGE (a) MERGE (b) MERGE (a)-[:R]->(b) MERGE (b)-[:R]->(a)
+ * In case the graph is empty at the time the algebraic expression is constructed
+ * the operand `R` is the zero matrix, but when the last merge is evaluated `R` is created. */
+void AlgebraicExpression_SyncOperands(AlgebraicExpression *ae) {
+	assert(ae);
+	Graph *g = QueryCtx_GetGraph();
+	const GraphContext *gc = QueryCtx_GetGraphCtx();
+
+	for(uint i = 0; i < ae->operand_count; i++) {
+		AlgebraicExpressionOperand *operand = ae->operands + i;
+		const char *label = operand->label;
+		Schema *s = NULL;
+
+		// Intermidate matrix, e.g. [:A|B]
+		if(operand->free) {
+			GrB_Index dim = Graph_RequiredMatrixDim(g);
+			// Ok to simply resize.
+			GxB_Matrix_resize(operand->operand, dim, dim);
+			continue;
+		}
+
+		if(operand->diagonal) {
+			// Operand represents a node.
+			s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
```

Comment:
`GraphContext_GetSchema` (called here and at 718) rely on strcmp loops, but given that I'd expect a very small N for schema count, that doesn't necessarily need to be optimized. Still, we could cache the ID on the `AlgebraicExpressionOperand` with or as a replacement to `operand->label`, and only make this call while the schema doesn't yet exist.

Reply:
Yeah that would be an optimisation. we can also switch to rax.